cpp/src/arrow/io/file.cc

```diff
@@ -378,6 +378,77 @@ Status FileOutputStream::Write(const void* data, int64_t length) {
 
 int FileOutputStream::file_descriptor() const { return impl_->fd(); }
 
+// ----------------------------------------------------------------------
+// DirectFileOutputStream, change the Open, Write and Close methods from FileOutputStream
+// Uses DirectIO for writes. Will only write out things in 4096 byte blocks. Buffers leftover bytes
+// in an internal data structure, which will be padded to 4096 bytes and flushed upon call to close.
+
+class DirectFileOutputStream::DirectFileOutputStreamImpl : public OSFile {
+ public:
+  Status Open(const std::string& path, bool append) {
+    const bool truncate = !append;
+    return OpenWritable(path, truncate, append, true /* write_only */, true);
+  }
+  Status Open(int fd) { return OpenWritable(fd); }
+};
+
+DirectFileOutputStream::DirectFileOutputStream() { 
+  uintptr_t mask = (uintptr_t)(4095);
+  uint8_t *mem = static_cast<uint8_t *>(malloc(4096 + 4095));
+  cached_data = reinterpret_cast<uint8_t *>( reinterpret_cast<uintptr_t>(mem+4095) & ~(mask));
```

Comment:
`posix_memalign`?

Actually, I'm not sure if an internal buffer is beneficial. AFAIK, direct IO is only for advanced apps like DMBS which implements their own buffer management and dislikes additional buffering (e.g., OS page cache). Looks to me Arrow as a library should not change this behaviour. Another issue is that we'd better leave IO alignment to the apps, as it depends heavily on OS and hardware, hardcode to 4096 implicitly is not good.

Reply:
Also, the biggest advantage of direct IO is to transfer data directly between the user mode buffer and disk controller. This additional buffering may render the advantage useless IIUC.

Reply:
Indeed, the `open` man page says this about `O_DIRECT`:
> Try to minimize cache effects of the I/O to and from this file.  In general this will degrade performance, but it is useful in special situations, such as when applications do their own caching.

@marsupialtail Are you seeing any actual benefits from this? In which situations?

Reply:
@marsupialtail this should maybe be linked against the JIRA https://issues.apache.org/jira/browse/ARROW-14635 which explains the goals a bit better.  The problem we are trying to solve here is not a performance problem. Instead, we are trying to keep large writes out of the OS cache.

An example use case is converting a large (~400GB) CSV dataset to parquet (maybe 100GB).  Another example might be repartitioning a large parquet dataset.  In both cases the user is unlikely to turn around and read the entire dataset and so there is no need to put the entire dataset into the OS cache, and this has detrimental effects.

What I've seen happen in these cases is that all of the data being written is stored in the page cache and writes return immediately (after the memcpy into the OS cache and before the data is persisted to disk).  Writes don't actually become blocking until the OS cache fills up.  The OS cache will use all available RAM and then the system grinds to a halt.  Furthermore, in the face of such memory pressure, even mildly inactive processes end up swapping, which increases the disk pressure.

As I mention in the JIRA, direct I/O is only one way to solve this problem.  Another way is by using sync writes and aggressively marking pages as DONTNEED.  @save-buffer did an experiment a while back and found that the sync/fadvise approach was considerably slower than the direct I/O approach (though I don't recall if this experiment involved caching on the direct I/O approach or not) and so we opted to go with direct I/O.

Reply:
I have trouble believing that `posix_fadvise` is really that much slower. You can also choose whatever granularity for `posix_fadvise` calls if you want to amortize the system call cost.

Reply:
And `posix_fadvise` is a much more generally applicable solution than `O_DIRECT`, which might very well be detrimental in many use cases.

Reply:
Will produce some benchmark experiments that compare the posix_fadvise with O_DIRECT and link here.

Reply:
For the record, when compiling your example I get:
```
fadvise.cpp: In function ‘int main()’:
fadvise.cpp:18:32: warning: integer overflow in expression of type ‘int’ results in ‘-2147483648’ [-Woverflow]
   18 |         posix_fadvise(fd, 0, N * SIZE, POSIX_FADV_DONTNEED);
      |                                ^
```

Reply:
Also, the "nothing" version is much faster here than the "direct" version (the former writes at ~300 MB/s, the latter at ~100 MB/s).

Reply:
this is valid. Let me fix this

Reply:
By the way:

> With O_DIRECT no page cache mem increase but the other two ways of writing caused 10GB page cache usage

I don't think this is conclusive. There is no reason for the OS to release some page cache is there's no memory pressure. I think a more sophisticated benchmark is needed, but I'm not sure what to suggest.

Reply:
[directio.zip](https://github.com/apache/arrow/files/9143508/directio.zip)
OK. I updated it to avoid overflow. The naive posix_fadvise approach (where you call this before ALL the writes) still doesn't work in controlling page cache. (fadvise.cpp)

However I discovered if you call posix_fadvise *every* time before a write on the *entire* region you intend to write, it works in controlling the page cache, despite being 50% slower than O_DIRECT. This is fadvise1.cpp. If you call posix_fadvise every time only on the next region you intend to write, it still doesn't control the page cache.

Bottom line is posix_fadvise probably works if we manage to guess at producing the right behavior, but it is a bit slower.

Reply:
That being said...I wonder how hard it would be to offer both?  I suppose if one is a clear winner we shouldn't burden users with the choice but, if there are pros/cons, or different environments support different options, we should probably offer the choice.

Reply:
> Also note that any use of synchronous writes means your main thread may block when writing

Fortunately the dataset writer has already been written with blocking writes in mind :wink: .  Writes are issued on the I/O thread pool and a backpressure mechanism is in place.  In fact, the inability of that backpressure mechanism to prevent swapping/thrashing is exactly what motivated this JIRA in the first place.

> My intuition is just that you may get glass jaws in real-world performance that you won't observe using a simple micro-benchmark on an otherwise local storage.

I'm not sure what you mean by this.  Is this a concern regarding direct vs fadvise or a concern regarding sync vs async?

Reply:
> I'm not sure what you mean by this. Is this a concern regarding direct vs fadvise or a concern regarding sync vs async?

Both. TBH, it's an entirely unsubstantiated intuition, but taking full control of how and when writes happen (as opposed to letting the kernel decide) means you'll face the same set of issues that kernel developers have dealt with and solved other the years.

Reply:
[directio.zip](https://github.com/apache/arrow/files/9145458/directio.zip)
Please disregard the previous uploaded code. @westonpace helped me get these new benchmarks. Got fadvise working. You have to use fadvise after the write sys call. Unaligned/aligned doesn't make too much of a difference for the O_DIRECT since memcpy compared to NVMe is cheap. Doing nothing (nothing.cpp) is around same speed as O_DIRECT if you call sync at the end. fadvise is marginally slower (by about 15%).

Reply:
So I am going to close this PR and make a new one where we just use fadvise.

Reply:
What about adding these benchmarks to Arrow code base?